providers/apis/dydx/utils.go

```diff
@@ -112,128 +144,190 @@ func CreateTickerFromMarket(market dydxtypes.MarketParam) (mmtypes.Ticker, error
-	paths := make([]mmtypes.Path, 0)
-	providers := make([]mmtypes.ProviderConfig, 0)
+) (mmtypes.Paths, mmtypes.Providers) {
+	var (
+		paths     []mmtypes.Path
+		providers []mmtypes.ProviderConfig
+		seen      = make(map[dydxtypes.ExchangeMarketConfigJson]struct{})
+	)
 
-	seen := make(map[dydxtypes.ExchangeMarketConfigJson]struct{})
 	for _, cfg := range config.Exchanges {
+		// Ignore duplicates.
 		if _, ok := seen[cfg]; ok {
 			continue
 		}
+		seen[cfg] = struct{}{}
+
+		// This means we have seen an exchange that slinky cannot support.
+		exchangeNames, ok := providerMapping[cfg.ExchangeName]
+		if !ok {
+			continue
+		}
 
 		var (
-			operations     []mmtypes.Operation
-			offChainTicker = ConvertDenomByProvider(cfg.Ticker, cfg.ExchangeName)
+			exchangePaths []mmtypes.Path
+			err           error
+			addProviders  = true
 		)
-
+		// Determine the relevant operations and provider configs based on the exchange config.
 		switch {
 		case len(cfg.AdjustByMarket) == 0 && !cfg.Invert:
-			// Direct conversion.
-			operation := mmtypes.Operation{
-				CurrencyPair: ticker.CurrencyPair,
-				Provider:     cfg.ExchangeName,
-				Invert:       false,
-			}
-			providerCfg := mmtypes.ProviderConfig{
-				Name:           cfg.ExchangeName,
-				OffChainTicker: offChainTicker,
-			}
-
-			operations = append(operations, operation)
-			providers = append(providers, providerCfg)
+			exchangePaths = DirectConversion(ticker, exchangeNames)
 		case len(cfg.AdjustByMarket) == 0 && cfg.Invert:
-			// Direct conversion with inverted price.
-			operation := mmtypes.Operation{
-				CurrencyPair: ticker.CurrencyPair,
-				Provider:     cfg.ExchangeName,
-				Invert:       true,
-			}
-
-			providerCfg := mmtypes.ProviderConfig{
-				Name:           cfg.ExchangeName,
-				OffChainTicker: offChainTicker,
-			}
-
-			operations = append(operations, operation)
-			providers = append(providers, providerCfg)
+			exchangePaths = InvertedConversion(ticker, exchangeNames)
 		case len(cfg.AdjustByMarket) > 0 && !cfg.Invert:
-			// Indirect conversion with index price. This is effectively a conversion
-			// from the base currency to the quote currency via the index currency.
-			// Ex. BTC/USD via BTC/USDT and USDT/USD.
-			first := mmtypes.Operation{
-				CurrencyPair: ticker.CurrencyPair,
-				Provider:     cfg.ExchangeName,
-				Invert:       false,
-			}
+			exchangePaths, err = IndirectConversion(ticker, cfg, exchangeNames)
+		case len(cfg.AdjustByMarket) > 0 && cfg.Invert:
+			exchangePaths, err = IndirectInvertedConversion(cfg, exchangeNames)
+			addProviders = false
+		}
 
-			cp, err := CreateCurrencyPairFromPair(cfg.AdjustByMarket)
-			if err != nil {
-				return mmtypes.Paths{}, mmtypes.Providers{}, err
-			}
-			second := mmtypes.Operation{
-				CurrencyPair: cp,
-				Provider:     mmtypes.IndexPrice,
-				Invert:       false,
+		// We passively ignore errors here, as we don't want to fail the entire conversion.
+		if err != nil {
+			continue
```

Comment:
log error?

Reply:
actually this might be a more difficult since i would have to attach the logger to the API handler, I'd prefer to have this be an exportable helper function in the case that dYdX wants to use it to test things.